confidant/services/certificate_authority/customca.py

```diff
@@ -0,0 +1,225 @@
+"""
+Custom Certificate Authority Module
+
+This module provides functionality for managing Custom Certificate
+Authorities (CAs) supplied by the user. It supports the signing of
+certificates for Certificate Signing Requests (CSRs).
+
+Note: This module does not provide functionality for generating keys or CSRs.
+"""
+
+import logging
+from datetime import datetime, timedelta, timezone
+
+from cerberus import Validator
+from cryptography import x509
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.hashes import SHA256
+
+from confidant import settings
+from confidant.services.certificate_authority.certificateauthoritybase import (
+    CertificateAuthorityBase,
+    CertificateAuthorityNotFoundError,
+)
+from confidant.settings import (
+    CUSTOM_CA_ACTIVE_KEYS,
+    CUSTOM_CERTIFICATE_AUTHORITIES,
+)
+
+logger = logging.getLogger(__name__)
+
+CUSTOM_CA_SCHEMA = {
+    "rootcrt": {"type": "string", "required": True, "nullable": True},
+    "crt": {"type": "string", "required": True},
+    "key": {"type": "string", "required": True},
+    "passphrase": {"type": "string", "required": True, "nullable": True},
+    "kid": {"type": "string", "required": True},
+}
+
+
+class CustomCertificateAuthority(CertificateAuthorityBase):
+    """Custom Certificate Authority
+
+    Args:
+        CertificateAuthorityBase (_type_): Base class for Certificate Authority
+    """
+
+    def __init__(self, ca_env: str):
+        self.ca_id = ca_env
+        self.active_ca_id = None
+        self.ca_json = self._get_ca_in_json(ca_env)
+        self.ca_certificate = self._load_ca_certificate(self.ca_json)
+        self.root_ca_certificate = self._load_rootca_certificate(self.ca_json)
+        self.ca_private_key = self._load_private_key(self.ca_json)
+        self.ca_chain = self._load_ca_chain()
+        self.settings = settings.CUSTOM_CA_SETTINGS
+
+    def _get_ca_in_json(self, ca_env: str):
+        if (
+            not CUSTOM_CERTIFICATE_AUTHORITIES
+            or ca_env not in CUSTOM_CERTIFICATE_AUTHORITIES
+        ):
+            logger.error("Custom CA %s not found", ca_env)
+            raise CertificateAuthorityNotFoundError(
+                f"Custom CA {ca_env} not found"
+            )
+        if not CUSTOM_CA_ACTIVE_KEYS or ca_env not in CUSTOM_CA_ACTIVE_KEYS:
+            logger.error("Custom CA %s has no active keys", ca_env)
+            raise CertificateAuthorityNotFoundError(
+                f"Custom CA {ca_env} has no active keys"
+            )
+        validator = Validator(CUSTOM_CA_SCHEMA)
+        active_ca_id = CUSTOM_CA_ACTIVE_KEYS[ca_env]
+        self.active_ca_id = active_ca_id
+        active_ca = [
+            ca
+            for ca in CUSTOM_CERTIFICATE_AUTHORITIES[ca_env]
+            if validator.validate(ca) and ca["kid"] == active_ca_id
```

Comment:
if validator.validate(ca) return False, will we still log Custom CA %s has no active keys?